### PR TITLE
[resharding] Handle case of restarting node during resharding catchup

### PR DESCRIPTION
**Pull request by shreyan-gupta**
This PR is a fix for the issue where when we restart a node doing resharding and in the catchup phase.

High level issue: When we restart a node in the epoch when resharding in happening, what happens is we go through the whole state_sync process from the beginning which includes resharding. Once building of the child trie is completed, we then do a catchup, apply the split state to the child tried and delete the split state changes for resharding [here](https://github.com/near/nearcore/blob/e00dcaa72cfed35831b1e72760d21bb8152f1049/chain/chain/src/chain_update.rs#L399).

Zulip thread link: https://near.zulipchat.com/#narrow/stream/308695-pagoda.2Fprivate/topic/Problems.20after.20resharding.20restart/near/421312417

This is the implementation of Option 1 in the thread.

The key idea here is to not delete the `DBCol::StateChangesForSplitStates` during catchup of individual blocks but rather at the end of the catchup phase. This implies, in case we restart the node in the middle the catchup, we would still have all the split state information in `DBCol::StateChangesForSplitStates` and it wouldn't have been deleted.

TODO: Testing on mocknet.

### CODE REVIEW

**Review by posvyatokum — COMMENTED** [review]
Looks good as a fix to the original issue.
I want to wait for mocknet testing to complete to see that any new issues do not occur during catchup after resharding.

As I understand, in current solution restart during catchup would mean that we start catchup from "scratch" (sync hash) again. And this catchup can take ~hour. I don't see it as a problem, I just want to improve release notes, so that this behaviour does not come as a surprise to validators. 

So, no notes on code, will approve after testing.

**Comment by shreyan-gupta** [comment]
> As I understand, in current solution restart during catchup would mean that we start catchup from "scratch" (sync hash) again.

Yes, that's correct! It's not ideal but probably the easiest to go by as once the catchup phase starts, the child tries are modified and can't be directly used after restart.

**Review by wacban — COMMENTED** [review]
LGTM but I'll let @posvyatokum approve once mocknet test is finished. 

Is there an easy way to unit / integration / nayduck test that restart in the middle of catchup works?

**Comment by marcelo-gonzalez** [comment]
Looks fine to me, sorry for not commenting on this earlier. Wasn't sure if I was the right person to approve this. However I did test this on mocknet and it seems that the bug where restarting after resharding is completed leads to the node not being able to progress into the new post-resharding epoch still exists. I'll investigate more and post some more info on what is happening. But @wacban if this change looks good to you, feel free to approve, since it doesn't seem to me like this makes anything worse

**Comment by marcelo-gonzalez** [comment]
ah actually it may have been a different failure. This time around after I restarted them, they failed because the resharding didn't finish, probably due to the changes here? So that's actually a different kind of failure, but I still need to investigate

**Comment by marcelo-gonzalez** [comment]
ok after looking at it more, I'm pretty sure this PR doesn't quite fix the bug. Will continue to investigate it

**Comment by marcelo-gonzalez** [comment]
ok sorry for all the noise... Actually I think I was wrong before, and this PR might actually fix it. It looks like before this PR the problem we're running into is that after restarting when the node is inside `Chain::catchup_blocks_step()`, [this apply chunk preprocessing call](https://github.com/near/nearcore/blob/61fef83d3bfde2b4ad3ff3e0e3306b9970535144/chain/chain/src/chain.rs#L2809) fails with an error:

```
2024-02-23T09:28:08.765068Z ERROR catchup: client: Some(AccountId("...")) Error occurred during catchup for the next epoch: Chain(DBNotFoundErr("CONSOLIDATED STATE CHANGES: 2RpVqkzTAaXpc1MbNkTdDd458xDQybMMfFRv9RZxuDCc:0"))
```

which comes from `Chain::get_update_shard_job()` [here](https://github.com/near/nearcore/blob/61fef83d3bfde2b4ad3ff3e0e3306b9970535144/chain/chain/src/chain.rs#L3440), and is caused by the deletion in the `StateChangesForSplitStates` by the previous run of neard.

So I think what happened is that I noticed that the chain stalled when getting to the new epoch and I stopped the nodes thinking that it had failed the same way as before. But actually it was stalled because too many validators who had been restarted were finishing the catchup, and if I had let it run it would have recovered (and this wouldn't be expected to be as big of a problem on mainnet with longer epochs than the ~4 hours I was running this with). This is supported by some extra debug log lines that I added to `Chain::catchup_blocks_step()` right before the apply chunk preprocessing is started. Before this PR, after restarting a node in the middle of catchup, we saw only one log line corresponding to the first block in the epoch, and then no more log lines after the error I printed above shows up. But after this PR, that error didn't show up and we got many debug log lines showing that catchup was proceeding normally.

So I'm going to run it one more time without stopping it right away when the chain stalls, and if the chain eventually progresses, then we'll see that this PR is indeed a mitigation of the issue (and that the stalling is just due to the shorter epoch length I'm working with).

I think eventually implementing @shreyan-gupta 's Option 2 would be pretty nice, since it seems like a shame to have to apply all these chunks all over again when that work has already been done pre-restart. But it looks to me like this PR is at least a pretty good mitigation in that it fixes the fact that your node will never recover after a restart.

After I re-run this test and verify that this worked, I'll go ahead and approve this PR even though I am not a resharding person :), since in that case I'll have p good confidence that it is good. But @wacban feel free to look more at it and approve/comment after I do so

**Comment by wacban** [comment]
LGTM! Ship it :)

**Review by marcelo-gonzalez — APPROVED** [review]

**Pull request merged**
